tmc/pkg/server/controllers.go

```diff
@@ -0,0 +1,406 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"fmt"
+	_ "net/http/pprof"
+
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	kcpkubernetesclientset "github.com/kcp-dev/client-go/kubernetes"
+
+	kcpapiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/kcp/clientset/versioned"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource"
+	schedulinglocationstatus "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/location"
+	schedulingplacement "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/placement"
+	workloadsapiexport "github.com/kcp-dev/kcp/pkg/reconciler/workload/apiexport"
+	workloadsapiexportcreate "github.com/kcp-dev/kcp/pkg/reconciler/workload/apiexportcreate"
+	"github.com/kcp-dev/kcp/pkg/reconciler/workload/heartbeat"
+	workloadnamespace "github.com/kcp-dev/kcp/pkg/reconciler/workload/namespace"
+	workloadplacement "github.com/kcp-dev/kcp/pkg/reconciler/workload/placement"
+	workloadresource "github.com/kcp-dev/kcp/pkg/reconciler/workload/resource"
+	synctargetcontroller "github.com/kcp-dev/kcp/pkg/reconciler/workload/synctarget"
+	"github.com/kcp-dev/kcp/pkg/reconciler/workload/synctargetexports"
+)
+
+func postStartHookName(controllerName string) string {
+	return fmt.Sprintf("kcp-tmc-start-%s", controllerName)
+}
+
+func (s *Server) installWorkloadResourceScheduler(ctx context.Context, config *rest.Config) error {
+	config = rest.CopyConfig(config)
+	config = rest.AddUserAgent(config, workloadresource.ControllerName)
+	dynamicClusterClient, err := kcpdynamic.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	resourceScheduler, err := workloadresource.NewController(
+		dynamicClusterClient,
+		s.Core.DiscoveringDynamicSharedInformerFactory,
+		s.Core.KcpSharedInformerFactory.Workload().V1alpha1().SyncTargets(),
+		s.Core.KubeSharedInformerFactory.Core().V1().Namespaces(),
+		s.Core.KcpSharedInformerFactory.Scheduling().V1alpha1().Placements(),
+	)
+	if err != nil {
+		return err
+	}
+
+	return s.Core.AddPostStartHook(postStartHookName(workloadresource.ControllerName), func(hookContext genericapiserver.PostStartHookContext) error {
+		logger := klog.FromContext(ctx).WithValues("postStartHook", postStartHookName(workloadresource.ControllerName))
+		if err := s.Core.WaitForSync(hookContext.StopCh); err != nil {
+			logger.Error(err, "failed to finish post-start-hook")
+			return nil // don't klog.Fatal. This only happens when context is cancelled.
+		}
+
+		go resourceScheduler.Start(ctx, 2)
```

Comment:
I see `2` is passed in a number of places, should we make this a constant instead?

Reply:
@ncdc has a PR for that I believe

Reply:
I do 😄